src/Domain/Services/ExtendRestApi.php

```diff
@@ -0,0 +1,129 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+use Throwable;
+
+/**
+ * Service class to provide utility functions to extend REST API.
+ */
+class ExtendRestApi {
+	/**
+	 * Valid endpoints to extend
+	 *
+	 * @var array
+	 */
+	private $endpoints = [ 'item' ];
+
+	/**
+	 * An endpoint that validates registration method call
+	 *
+	 * @param string   $endpoint The endpoint to extend.
+	 * @param string   $namespace Plugin namespace.
+	 * @param callable $schema_callback Callback executed to add schema data.
+	 * @param callable $data_callback Callback executed to add endpoint data.
+	 */
+	private function validate_endpoint_data( $endpoint, $namespace, $schema_callback, $data_callback ) {
+		if ( ! is_string( $namespace ) ) {
+			$this->throw_exception(
+				'You must provide a plugin namespace when extending a Store REST endpoint.'
+			);
+		}
+
+		if ( ! is_string( $endpoint ) || ! in_array( $endpoint, $this->endpoints, true ) ) {
+			$this->throw_exception(
+				'You must provide a valid Store REST endpoint to extend, valid endpoints are: ' . implode( ', ', $this->endpoints )
+			);
+		}
+
+		if ( ! is_callable( $schema_callback ) ) {
+			$this->throw_exception(
+				'$schema_callback must be a callable function.'
+			);
+		}
+
+		if ( ! is_callable( $data_callback ) ) {
+			$this->throw_exception(
+				'$data_callback must be a callable function.'
+			);
+		}
+	}
+
+	/**
+	 * Extends the items in store/cart endpoint
+	 *
+	 * @param string   $namespace Plugin namespace.
+	 * @param callable $schema_callback Callback executed to add schema data.
+	 * @param callable $data_callback Callback executed to add endpoint data, gets cart item data as param.
+	 */
+	public function register_cart_item_endpoint_data( $namespace, $schema_callback, $data_callback ) {
```

Comment:
`register_cart_item_endpoint_data` is the first function we're going to implement, future ones are going to be things like `register_shipping_endpoint_data` `register_product_endpoint_data` `register_cart_endpoint_data` `register_checkout_endpoint_data`

Reply:
Instead of creating individual method names for each registration point, what about having something like:

```php
public function register_endpoint_data( $identifier, $namespace, $schema_callback, $data_callback ) { /**/ }
```

`$identifier` will be something like `cart_item`, `shipping` etc and that would be a part of the private `$valid_identifiers` property which can be used for validation:

```php
private $valid_identifiers = [ 'cart-items', 'shipping' ];
```

As far as identifiers, I would standardize around the ones used in the [`RoutesController`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/edda1fa24e4f0437737d8f738b9dce4b9505cc31/src/StoreApi/RoutesController.php#L69-L93) (similar to what is also used in the [`SchemaController`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/StoreApi/SchemaController.php).

As an aside, these identifiers should ideally be constants to reduce potential for human error and also extensions can use the constants. I would like to see the constants added in the concrete classes. So for example [`Automattic\WooCommerce\Blocks\StoreApi\Routes\CartItems`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/StoreApi/Routes/CartItems.php) would have this defined in the class:

```php
const IDENTIFIER = 'cart-items'
```

Then in [`RoutesController`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/edda1fa24e4f0437737d8f738b9dce4b9505cc31/src/StoreApi/RoutesController.php#L68-L94) you'd end up with something like this for each item in the `routes` array:

```php
$this->routes = [
     Routes\Cart::IDENTIFIER => new Routes\CartItems( $this->schemas->get( Schemas\CartItemSchema::IDENTIFIER ) ),
];
```
(as you can tell you'd do something similar for schemas).

Then extensions can reference the identifiers without risk of human error (and IDE's will appropriately enhance references etc) by the class constant when registering their extra data via the new interface.

This class constant implementation could be done in a PR and merged separately from this pull request and brought in to it after merge (which would make the review easier).

Reply:
> Instead of creating individual method names for each registration point

That was my original goal, and that's how `validate_endpoint_data` used to work (it was named `register_endpoint_data`), but because I was using filters, there's the limitation of passed arguments.
So Cart Item filter might pass 2 arguments, and shipping might pass just 1 or 3.
If you don't specify that value in the `add_filter` call, you would get a fatal error, so each endpoint wrapper was needed.
If we move away from filters, I will try implementing this.